subsys/logging/log_backend_cavs_hda.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "toolchain/gcc.h"
+#include <spinlock.h>
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <logging/log_backend_std.h>
+#include <logging/log_backend_cavs_hda.h>
+#include <drivers/dma.h>
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_CAVS_HDA_OUTPUT_DEFAULT;
+static const struct device *hda_log_dev;
+static uint32_t hda_log_chan;
+
+/* Trace output goes in 64 byte chunks with a 4-byte header. HDA however
+ * requires 128 byte aligned data and 128 byte aligned transfers.
+ * So regardless of the log output we *always* send 128 bytes. The other
+ * side then can interpret it as needed.
+ */
+static __aligned(128) uint8_t hda_log_buf[CONFIG_LOG_BACKEND_CAVS_HDA_SIZE];
+static volatile uint32_t hda_log_buffered;
+static volatile bool initialized, panic_mode;
+static atomic_t hda_log_lock;
+static struct k_timer hda_log_timer;
+static cavs_hda_log_hook_t hook;
+
+static void hda_log_flush(void)
+{
+	/* Flushing is a best effort, if we fail now we can catch it a little
+	 * later
+	 */
+	if (!atomic_cas(&hda_log_lock, 0, 1)) {
+		return;
+	}
+
+	uint32_t nearest128 = hda_log_buffered & ~((128) - 1);
+
+	if (nearest128 > 0) {
+		hda_log_buffered = hda_log_buffered - nearest128;
+
+		dma_reload(hda_log_dev, hda_log_chan, 0, 0, nearest128);
+
+		if (hook != NULL) {
+			hook(nearest128);
+		}
+	}
+
+	atomic_set(&hda_log_lock, 0);
+}
+
+
+static int hda_log_out(uint8_t *data, size_t length, void *ctx)
+{
+	int ret;
+	struct dma_status status;
+
+	while (!atomic_cas(&hda_log_lock, 0, 1)) {
+	}
+
+	/* Defaults when DMA not yet initialized */
+	uint32_t dma_free = sizeof(hda_log_buf);
+	uint32_t write_pos = 0;
+
+	if (initialized) {
+		ret = dma_get_status(hda_log_dev, hda_log_chan, &status);
+		if (ret != 0) {
+			ret = length;
+			goto exit;
```

Comment:
Don't get this bit.  This is treating an error in dma_get_status (in a device that wasn't even initialized!) as... success and claiming to have accepted all the data correctly?

Reply:
OK.  But now I have another question: why is it valid to call dma_get_status() here if initialized is unset (which we now know means "the device hasn't been assigned an HDA channel by the host driver")?  Still not understanding.

Reply:
this is why dma_get_status was wrapped in if (initialized)